cpp/embeddings.cc

```diff
@@ -0,0 +1,42 @@
+#include <cstring>
+#include  <stdexcept>
+
+#include "embeddings.h"
+#include "finalfusion.h"
+
+Embeddings::Embeddings(std::string const &filename, bool const mmap) {
+  ff_embeddings embeddings;
+  if (mmap) {
+    embeddings = ff_mmap_embeddings(filename.c_str());
+  } else {
+    embeddings = ff_read_embeddings(filename.c_str());
+  }
+
+  if (embeddings == nullptr) {
+    throw std::runtime_error(std::string(ff_error()));
+  }
+  d_inner = embeddings;
+}
+
+Embeddings::~Embeddings() {
+  ff_free_embeddings(d_inner);
+}
+float *Embeddings::embeddingRaw(std::string const &word) {
```

Comment:
The motivation for this was to avoid a second heap allocation before being able to copy the embedding.

But I remember premature optimization being the root of all evil, so let's kick this one out until we profiled whether this can actually be a bottleneck.

Reply:
This is a shortcoming of C++'s `vector` (unless you use a custom allocator). E.g. Armadillo's vector type also provides a constructor that takes an existing pointer. Not sure about Eigen.

C++ 20 provides a `span` type. On pre-20 compilers the Microsoft GSL library also provides `span`.